feat: NetworkManager Preparation [MTT-5441]

NoelStephensUnity opened the pull request:
Extracting systems from NetworkManager to prepare for future epic work.

[MTT-5441](https://jira.unity3d.com/browse/MTT-5441)

This includes a minor fix related to #2466 that seemed better to apply to this feat than to make the suggested changes in the PR.  This also includes an update to the DisconnectTests that broadens the coverage and includes an additional check to assure the server-side transport-client ids are cleaned up on the server-side.


## Testing and Documentation

- No tests have been added.
- Integration tests have been updated/modified.
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
update - NetworkConnectionManager

First pass potential way to extract the connection management into a separate NetworkConnectionManager.

Passes all integration tests but the connection approval timeout (pending clients).

NoelStephensUnity committed:
update

Some remaining adjustments to assure all tests are passing.
Still has some aspects not migrated over, but the majority of connection management is migrated into its own class.

NoelStephensUnity committed:
Merge branch 'develop' into experimental/networkmanager-systems

NoelStephensUnity committed:
test

Increased initial wait period as running the ConnectionApprovalTimeoutTests by themselves as a stand alone runtime build would yield a failure due to systems still initializing. (doesn't impact the test other than when only building the individual test as a runtime stand alone build).

NoelStephensUnity committed:
update

Removed as much of the connection related code from NetworkManager as possible for this pass.
Migrated SendMessage into ConnectionManager.
Migrated the remainder of the connection approval related code into ConnectionManager.
Created a NetworkMetricsManager where NetworkMetrics now lives.
Migrated all of the connection related EarlyUpdate code from NetworkManager into ConnetionManager.

All tests passed with this update applied.

NoelStephensUnity committed:
update

Cleaning up NetworkClient so we don't need to create an instance to reset it.
Minor style clean up.

NoelStephensUnity committed:
update

Getting rid of the Player object requirement to find objects owned by the client.
One could theoretically connect and never have an associated player object, but still own NetworkObjects.

NoelStephensUnity committed:
Merge branch 'develop' into experimental/networkmanager-systems

NoelStephensUnity committed:
update

Adding the integration test fix from PR-2411
Updating NetworkManager to the NetworkClient adjustments.

NoelStephensUnity committed:
Merge branch 'develop' into feat/networkconnectionmanager-extracted-from-networkmanager

NoelStephensUnity committed:
Merge Update

Handling diffs when merging develop.

NoelStephensUnity committed:
merge update

Some updates for minor diffs when merging develop into this branch.

NoelStephensUnity committed:
update

minor updates

NoelStephensUnity committed:
update

NoelStephensUnity committed:
update

organizing a bit and still extracting things.

NoelStephensUnity committed:
update

A bit more reorganizing

NoelStephensUnity committed:
style

Removing whitespaces

NoelStephensUnity committed:
update

Fixing reference to ClientIdToTransportId for tools

NoelStephensUnity committed:
style

removing unused namespace

NoelStephensUnity committed:
update

reverting some changes that failed the package validation suite.
Two were technically valid (but not something that would break users)
One I think the package validation suite made a mistake:
public IReadOnlyDictionary<ulong, PendingClient> PendingClients => ConnectionManager.PendingClients;

Yields the same result (API wise) as:
public readonly Dictionary<ulong, PendingClient> PendingClients = new Dictionary<ulong, PendingClient>();

NoelStephensUnity committed:
update

DisconnectClient(ulong clientId) was deleted by accident.

NoelStephensUnity committed:
test fix

With the updates, found a test that needed some touch up work and sort of surprised it worked in the first place.

NoelStephensUnity committed:
Merge branch 'develop' into feat/networkmanager-preparation-mtt-5441

NoelStephensUnity committed:
update (merged develop)

Modifications from the recent updates pushed to develop.

NoelStephensUnity committed:
update - no networkmanger updates

This includes the changes required to migrate all NetworkUpdateLoop updates outside of NetworkManager.
There is still some additional clean up to do with properties, removing #regions, need to review over the changes made a bit longer, and there is some clean up that could occur within NetworkConnectionManager.

NoelStephensUnity committed:
update

Cleaning up NetworkConnectionManager,
Making adjustments to NetworkManager

NoelStephensUnity committed:
update - tools

Fixing issue with the MetricHooks when MULTIPLAYER_TOOLS is defined.

NoelStephensUnity committed:
style

adding a comment to clarify what MessagingSystem.ClientConnected actually does.

NoelStephensUnity committed:
update

Some NetworkTimeSystem adjustments... not sure about the recent standards imports ordering issue...

NoelStephensUnity committed:
update

Fixing the import ordering... alphabetically...

NoelStephensUnity committed:
style

updating comments

NoelStephensUnity committed:
style

more comment updates

NoelStephensUnity committed:
fix

Fixing some typos.

NoelStephensUnity committed:
update

The problem with interfaces... the implemented methods have to be public.  :(
Adding public void NetworkUpdate(NetworkUpdateStage updateStage) back into NetworkManager to avoid the package validation check issue.
Removing the internal constructor from NetworkBehaviourUpdater and just exiting the constructor for testing purposes.

NoelStephensUnity committed:
Revert "update"

This reverts commit a5be41f891a2a0ff5c056b0bfc4f9da1d55a14fd.

NoelStephensUnity committed:
update - take 2

(without all the extra stuff)
The problem with interfaces... the implemented methods have to be public. :(
Adding public void NetworkUpdate(NetworkUpdateStage updateStage) back into NetworkManager to avoid the package validation check issue.
Removing the internal constructor from NetworkBehaviourUpdater and just exiting the constructor for testing purposes.

NoelStephensUnity committed:
update

The transport failure message changed slightly.
Consolidated the message generation.

NoelStephensUnity committed:
update

removing NetworkBehaviour constructor completely and using internal Initialize to avoid package validation crazy.

NoelStephensUnity committed:
update - clean up

Cleaning up some of the NetworkManager start code to keep a more standard pattern.
Migrated the host-client specific code into its own method to help with this.
Updated the transport failure network event handler to call shutdown internal directly if  it occurs when starting the NetworkManager or just do the normal shutdown with stopping message processing if it occurs during a network session.

NoelStephensUnity committed:
test - tools (temporary)

Temporary work around for a perplexing total bytes test.
Need to get with tools and Kitty about this one before proceeding.

NoelStephensUnity committed:
update

Fixed a missed migration issue where it was synchronizing time every tick and not at the defined timed synchronization frequency.
Reverted some changes made to TansportBytesMetricsTests that were perfectly fine (the time synchronization message being sent each tick was throwing it off on the total bytes sent).
Some minor clean up and additional comments.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Merge branch 'develop' into feat/networkmanager-preparation-mtt-5441

NoelStephensUnity committed:
update - merge diffs

Updating the minor merge differences.
Making minor modification to NetworkConnectionManager to prevent from trying to despawn a player if the player is already despawned.

NoelStephensUnity committed:
Merge branch 'develop' into feat/networkmanager-preparation-mtt-5441

NoelStephensUnity committed:
style

removing the UnitEngine namespace reference

NoelStephensUnity committed:
update

Migrating prefab specific code out of NetworkManager and into NetworkPrefabHandler.
Did a final clean up pass with NetworkManager.

NoelStephensUnity committed:
update

Migrating NetworkMetricsManager into Metrics.
Renamed NetworkManagerSender to DefaultMessageSender and migrated it into Messaging.

NoelStephensUnity committed:
update

Renaming NetworkTimeSystem.InitializeForUpdates to NetworkTimeSystem.Initialize to keep this method name consistent with the rest.

NoelStephensUnity committed:
style

Just updating the comments to reflect why the added method is there and why it is invoked by NetworkConnectionManager.

simon-lemay-unity commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkConnectionManager.cs:
```diff
@@ -0,0 +1,1095 @@
+        /// <summary>
+        /// The current host name we are connected to, used to validate certificate
+        /// </summary>
+        public string ConnectedHostname { get; private set; }
```
Might want to remove this entirely (and just have the property return `null` in `NetworkManager`). This isn't used by anything and bringing it along in new APIs could make it look like it is.

NoelStephensUnity replied:
Done! Just returning string.Empty now.

simon-lemay-unity replied:
I should have been clearer in my comment. I was actually recommending returning an empty string from the getter in `NetworkManager`, and simply not having the property in `NetworkConnectionManager`. I get that the property needs to stay in `NetworkManager` to avoid breaking the API, but it doesn't have to be in new APIs like `NetworkConnectionManager`, as this makes it look like the API is actually useful.

0xFA11 replied:
done as you suggested, see c8dd262

simon-lemay-unity reviewed (approved):
I don't have the broader context to provide a meaningful review of the overall changes, but transport changes look fine and this will get the server team's review out of the way. 😃

NoelStephensUnity committed:
update

NetworkConnectionManager.ConnectedHostname now returns string.Empty per suggestion since this is no longer used.

NoelStephensUnity committed:
style

Left out a "t" (typo).

NoelStephensUnity committed:
update

Moving ConnectionManager.Initialize(this) below the check for the parent.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkClient.cs:
```diff
@@ -24,13 +59,28 @@ public List<NetworkObject> OwnedObjects
+            ClientId = clientId;
+        }
+
+        public NetworkClient()
```
Is this actually valid? Creating a client that's neither server nor client?

Also, should these constructors be public? Do we actually expect users to create NetworkClients or should they be internal so we create them and then pass instances to users?

NoelStephensUnity replied:
The [NetworkClient ](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/develop/com.unity.netcode.gameobjects/Runtime/Connection/NetworkClient.cs)has actually been public and part of NGO from day 1 and even in MLAPI.
The default role state of the NetworkClient (for this pass) is to be in a "neither a server-host or a client" until the NetworkManager is started and can set the context of the NetworkClient (i.e. StartServer, StartHost, StartClient).  It is constructed at the time of the declaration, which makes accessing the related information always available (avoids having to check for other states... if you aren't a client or a server then you aren't connected...not that I am suggesting that we make that a way to determine connection state... but just that it sort of makes sense if a user were to check those properties).

_The intention was to migrate these specific properties away from a system and into something else...the NetworkClient seemed like a good place to keep these properties since it also holds the PlayerObject and such.  We could migrate these properties into NetworkConnectionManager, but that didn't seem right either.  Maybe the solution is to come up with another "object" that is internal and more easily changed... definitely worth discussing where we want this information being persisted.  I am not tied to having all of that information there, but it was a location that seemed to make sense at the time... but definitely could be easily changed before committing to these changes (for sure)._

Regarding the public constructor, the class has always been public which users could always construct a new instance (for what would be a good question)...if we plan on migrating to something completely different then perhaps we make all of the added properties internal and for the time being keep the NetworkClient public API exactly as is until we have a better place to store those properties?

ShadauxCat replied:
Could we make the new constructor internal, and make the existing one throw an exception or at least log an error? The actual class being public makes perfect sense, it's just the constructors being public that feels off to me. Not a hill I'll die on though.

NoelStephensUnity replied:
The new one needs to not throw a warning or error because it is used when instantiating the instance at declaration, but made the overloaded one internal.

0xFA11 replied:
we actually do not need to call `SetRole(false, false)` in the default public ctor because all properties are going to be initialized with their default values already (false, null, 0 etc.)

0xFA11 replied:
in fact, I will delete both ctors and call `SetRole(...)` from the original call site.

0xFA11 replied:
done, see c22dd73

NoelStephensUnity committed:
update

Some final re-organization and consolidation of the NetworkManager startup and shutdown sequence. All systems are spun up and shutdown in the order of their dependencies.
Migrated the player prefab registration out of NetworkManager into NetworkPrefabHandler.
Migrated transport instantiation out of NetworkManager into NetworkConnectionManager.

NoelStephensUnity committed:
update

last minor update, migrating the time system shutdown to be the last thing shutdown.

NoelStephensUnity committed:
fix

Assure TransportIdToClientId and ClientIdToTransportId will not throw an exception if the client or transport id does not exist. Includes a developer level warning if either are invoked when this condition occurs.

Calling NetworkConnectionManager.Disconnect directly in the approval timeout coroutine.

Assuring the transport-client id mapping tables on the server-side are cleaned up when a client is disconnected.

Consolidated where a client id is removed from the pending clients list (other than when a client is approved).

NoelStephensUnity committed:
test

updated the disconnecttests to leverage from the more current NetcodeIntegrationTest class, broadened the coverage, and added a check to assure the server-side cleans up the transport - client id mapping tables.

NoelStephensUnity committed:
update

Cleaning up NetworkClient a bit, adding public API comments, making the overloaded constructor internal, and changing the NetworkManager internal property to just the NetworkSpawnManager (what is only really needed).

NoelStephensUnity committed:
Merge branch 'develop' into feat/networkmanager-preparation-mtt-5441

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -38,231 +30,317 @@ public class NetworkManager : MonoBehaviour, INetworkUpdateSystem
-        private const float k_DefaultBufferSizeSec = 0.05f; // todo talk with UX/Product, find good default value for this
-
-        internal static string PrefabDebugHelper(NetworkPrefab networkPrefab)
+        // TODO 2023-Q2: Team discussion
```
these types of comments are promising a timeline inline in the codebase which we do not want to be liable for necessarily.

NoelStephensUnity replied:
They are to be removed after the final review before merging.  Just labeled in a way that was easy to find.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkConnectionManager.cs:
```diff
@@ -0,0 +1,1151 @@
+    /// - MessagingSystem updates
+    /// </summary>
+    // TODO 2023-Q2: Discuss what kind of public API exposure we want for this
+    public class NetworkConnectionManager : INetworkUpdateSystem
```
So thinking back to the original document for some of this was to have NetworkManager be the one that would call into the various components registered to it. Making this INetworkUpdateSystem sounds like it'll be registered with NetworkUpdateLoop instead. Is that just an intermediate step?

Main reason this concerns me as an intermediate step is that, when we create a different interface to be used for NetworkManager (since the way NetworkManager processes updates is a little different with the inclusion of dependencies within each stage), that'd be a breaking change from this. So if this goes into develop as it is, we'd be blocked from releasing anything until we get the next step done.

NoelStephensUnity replied:
Actually a good point... this should have had an internal INetworkUpdateSystem implementation that gets instantiated and hides the public NetworkUpdate method within NetworkConnectionManager itself.  That will avoid any future changes where perhaps the update is handled elsewhere (i.e. removing the implementation won't be considered a breaking change).
Regarding the order in which things are updated, they are still updated in the same order...just not all within NetworkManager.

0xFA11 replied:
Since we're not removing `NetworkManager.NetworkUpdate(...)` method yet, and `NetworkManager` is still the service locator for network systems, we can just simply call update on network systems (i.e. `NetworkConnectionManager`) from outside, from `NetworkManager.NetworkUpdate(...)` up until we make breaking changes.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -38,231 +30,317 @@ public class NetworkManager : MonoBehaviour, INetworkUpdateSystem
+        /// <summary>
+        /// Gets the networkId of the server
+        /// </summary>
+        private ulong m_ServerTransportId => ConnectionManager.ServerTransportId;
```
Style question... is `m_` still actually the appropriate prefix for a property? Feels a little odd to me.

NoelStephensUnity replied:
any private property requires the m_ prefix (as I understand it).

NoelStephensUnity committed:
update

Migrating INetworkUpdateSystem into internal classes to prevent from creating an additional public API method for NetworkConnectionManager and NetworkTimeSystem.

NoelStephensUnity committed:
update

Testing potential suggested approach for the NetworkManager.PendingClients issue.

NoelStephensUnity committed:
Revert "update"

This reverts commit 8965c44d4fd839b7ad5a3f6e4e00ecae304c752e.

NoelStephensUnity committed:
update

An alternate way to handle the pending clients issue.

NoelStephensUnity committed:
Merge branch 'develop' into feat/networkmanager-preparation-mtt-5441

NoelStephensUnity committed:
style

removing comments.

0xFA11 committed:
++

0xFA11 committed:
delete `NetworkManager.m_ServerTransportId`

0xFA11 committed:
minor `NetworkManager` refactor

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkClient.cs:
```diff
@@ -7,30 +7,95 @@ namespace Unity.Netcode
     /// </summary>
     public class NetworkClient
     {
+        /// <summary>
+        /// Returns true if the session instance is considered a server
+        /// </summary>
+        public bool IsServer { get; internal set; }
+
+        /// <summary>
+        /// Returns true if the session instance is considered a client
+        /// </summary>
+        public bool IsClient { get; internal set; }
+
+        /// <summary>
+        /// Returns true if the session instance is considered a host
+        /// </summary>
+        public bool IsHost => IsClient && IsServer;
+
+        /// <summary>
+        /// When true, the client is connected, approved, and synchronized with
+        /// the server.
+        /// </summary>
+        public bool IsConnected { get; internal set; }
+
+        /// <summary>
+        /// Is true when the client has been approved.
+        /// </summary>
+        public bool IsApproved { get; internal set; }
```
Even though I agree in principle that having these properties here make more sense, instead of NetworkManager, it adds more onto our public API surface/footprint.

I will revert this public API change in NetworkClient and use references to NetworkManager's.

We can (and will) revisit this when we get to breaking API changes later in the future.

NoelStephensUnity replied:
Be cautious with that... I would rather we make them all internal and expose them publicly via NetworkManager than to remove them all together.  Removing them completely will lead to a bunch of additional changes.

0xFA11 replied:
I saw that, I'll give it a try and pick between two.

0xFA11 replied:
made them internal as you suggested, see 0a8ccb2

0xFA11 committed:
delete `NetworkClient` ctors

0xFA11 committed:
make `NetworkClient` bool props internal

0xFA11 committed:
minor fix, do not skip Unity object equality check that also checks for object lifetime unlike ?. and/or ReferenceEquals

0xFA11 committed:
merge branch 'develop' into 'feat/networkmanager-preparation-mtt-5441'

0xFA11 committed:
delete `NetworkConnectionManager.ConnectedHostname`

0xFA11 committed:
minor `NetcodeIntegrationTest` fix

0xFA11 committed:
minor refactor

0xFA11 committed:
delete `NetworkConnectionManagerUpdater` and `NetworkTimeSystemUpdater` in favor of `NetworkManager`-driven `NetworkUpdate(updateStage)` into network systems
